helper/restore_helper.go

```diff
@@ -164,7 +191,23 @@ func doRestoreAgent() error {
 
 			filename := replaceContentInFilename(*dataFile, contentToRestore)
 			readers[contentToRestore], err = getRestoreDataReader(filename, segmentTOC[contentToRestore], oidList)
-
+			if readers[contentToRestore] != nil {
+				// NOTE: If we reach here with batches > 1, there will be
+				// *origSize / *destSize (N old segments / N new segments)
+				// readers + 1, which is presumably a small number, so we just
+				// defer the cleanup.
+				//
+				// The loops under are constructed in a way that needs to keep
+				// all readers open for the entire duration of restore (oid is
+				// in outer loop -- batches in inner loop, we'll need all
+				// readers for every outer loop iteration), so we can't properly
+				// close any of the readers until we restore every oid yet,
+				// unless The Big Refactoring will arrive.
+				defer func() {
+					readers[contentToRestore].waitForPlugin()
+					readers[contentToRestore].logPlugin()
+				}()
```

Comment:
we call waitForPlugin and logPlugin here and then on doRestoreAgent exit?

Reply:
these defer calls are for `singleDataFile`. And in the loop end `waitForPlugin` and `logPlugin` are called for `!*singleDataFile`. So they do not intersect.

Reply:
I mean `}()`